packages/app/src/authProviders.ts

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {
+  googleAuthApiRef,
+  gitlabAuthApiRef,
+  oktaAuthApiRef,
+  githubAuthApiRef,
+  commonProvider,
+  customProvider,
+  guestProvider,
+} from '@backstage/core';
+
+export const providers = [
+  {
+    id: 'google-auth-provider',
+    title: 'Google',
+    message: 'Sign In using Google',
+    apiRef: googleAuthApiRef,
+    provider: commonProvider,
```

Comment:
I want to be a bit mindful of what we're exporting from `@backstage/core`. I think this approach is definitely the best for any internal app, but since we're dealing with the public API I want to be a bit careful. My concern is the top-level `*Provider` exports, which are very specific to the `SignInPage`, but now a bit decoupled from it. Especially since the providers contain the component to be rendered, it can't really be reused in other sign-in pages.

One solution could be to simply add them as static members of the page, for example `SignInPage.commonProvider`, but extra static stuff on components is always a bit wonky.

I was thinking we could try to split the providers up into separate props, something like

```ts
type Props = {
  guest?: boolean
  custom?: boolean
  providers: { Component: ..., loader: ...}[]
  identityProviders: { id: ..., title: ..., apiRef: ... }[]
```

But that will inevitably lead to requests to be able to change the orders of them 😅 

Another option could be to merge all of those things into a array of union types, something like

```ts
type Props = {
  providers: ('guest' | 'custom' | { Component: ..., loader: ...} | { id: ..., title: ..., apiRef: ... })[]
}
```

I'm a bit torn between the array of union types and just going with static members, what do you think? :grin:

Reply:
I am wondering if there is a need for passing `providers: { Component: ..., loader: ...}[]` as a prop if we allow `guest` and `custom` to be booleans. My intention, when I added that prop, was to avoid conditions inside on the SignInPage component and keep this array of providers flexible.

I both examples (if I understand correctly) you can pass custom either an array of `identityProviders` (these will use commonProvider components),  `providers`, or combine both, right?

In this case, the first solution sounds a bit more explicit for me and it's probably easier to render those props, but I would keep all props optional. About the order, I would make a very bad assumption that most of the user will not have more that one 😄

Nevertheless, the second solution provides a custom order and it might be easier to maintain.

Reply:
Maybe just `providers: ('guest' | 'custom' | { id: ..., title: ..., apiRef: ... })[]` then? It does seem a bit extreme to be able to pass custom Component/loader pairs from the app, since the entire `SignInPage` is already pluggable in itself.

Regarding order... if the implementation is just a little bit more complex I think it's better to go with the single array tbh 😅 . I think it should be relatively simple either way, basically just mapping the string `guest` to the guestProvider you already have and so on.

When it comes to just having a single provider, I was actually thinking this `SignInPage` component would be renamed to `MultiProviderSignInPage`. Then we'd make a second one called `SingleProviderSignInPage` that has a bit nicer layout for using just a single provider, bigger centered box and such.

Reply:
The two different components might be actually a nice idea, but maybe we can keep that as a separate PR?

 In the case of having one `identityProvider` passed to the component title `Select a sign-in method ` is really redundant. About the styling, I was about to bring that up in later. The styling for that page can be improved by something like this (then we have responsive, centered cards of the same width and height):

```js
const useStyles = makeStyles({
  title: {
    textAlign: 'center',
  },
  container: {
    flexFlow: 'row wrap',
    justifyContent: 'center',
  },
  item: {
    display: 'flex',
    width: '100%',
    maxWidth: '400px',
  },
  infoCard: {
    flex: 1,
  }
});
```

Reply:
Oh yeah totally a separate PR for the page change indeed.

Regarding the style changes, maybe that can be a separate PR too? :grin:
Didn't try it out but I think I know where you're going with it and I agree. Please include screenshots of the change if you end up opening a PR!